optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,168 @@
+from collections import OrderedDict
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+import scipy
+
+import optuna
+from optuna import distributions
+from optuna import logging
+from optuna._transform import _SearchSpaceTransform
+from optuna.distributions import BaseDistribution
+from optuna.samplers import BaseSampler
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+_logger = logging.get_logger(__name__)
+
+_SUGGESTED_STATES = (TrialState.COMPLETE, TrialState.PRUNED)
+
+_NUMERICAL_DISTRIBUTIONS = (
+    distributions.UniformDistribution,
+    distributions.LogUniformDistribution,
+    distributions.DiscreteUniformDistribution,
+    distributions.IntUniformDistribution,
+    distributions.IntLogUniformDistribution,
+)
+
+
+class QMCSampler(BaseSampler):
+    def __init__(
+        self,
+        *,
+        seed: Optional[int] = None,
+        qmc_type: str = "sobol",
+        independent_sampler: Optional[BaseSampler] = None,
+        search_space: Optional[Dict[str, BaseDistribution]] = None,
```

Comment:
Not sure if this data type is appropriate for `search_space` option. In the case of `GridSampler`, it is `Dict[str, List[int]]`.

Reply:
It might be simple to infer the search space from the finished first trial as done in _infer_initial_search_space.

Reply:
Do you mean infering the search space from the trial that the user provided?
```suggestion
        initial_trial: Optional[FrozenTrial] = None,
```

Reply:
Thanks for the reply. If my understanding is wrong, please point it out.

First, let me check the current implementation: if the user specifies a search space in `__init__` of QMCSampler, the search space is always returned in `infer_relative_search_space`. If not, the search space is estimated by looking at the first trial in the `infer_relative_search_space` and returned.

If the search space is estimated after the first trial is completed, I think it would not be necessary to provide a way to specify the search space in `__init__` of QMCSampler.

Reply:
Sorry for the late reply!
> First, let me check the current implementation: if the user specifies a search space in __init__ of QMCSampler, the search space is always returned in infer_relative_search_space. If not, the search space is estimated by looking at the first trial in the infer_relative_search_space and returned.

Yes, that is correct.

> If the search space is estimated after the first trial is completed, I think it would not be necessary to provide a way to specify the search space in __init__ of QMCSampler.

I thought there might be users who want to use QMC sampler in a search space that is not one of the initial trials. For example, when they have enough resources, they might 1) run QMC sampler on the entire search space and look at HP importance and then 2) reduce the number of HPs and run QMC sampler again (just instead of grid search).

If the QMC sampler always samples from the same search space as the one of the initial trial, users have to create a new study every time they use the QMC sampler on different search spaces. If they do so, the information in the different studies would be hard to aggregate. (I am not sure if there is a good way in Optuna to import the trials from one study to another...)

Nevertheless, as this is rather an edge case, I agree to remove the `search_space` argument from the sampler, if you still believe that this simplification will be beneficial.